pymeasure/instruments/keysight/keysightN7776C.py

```diff
@@ -0,0 +1,256 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+import numpy as np
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import strict_discrete_set, strict_range
+
+WL_RANGE = [1480,1620]
+LOCK_PW = 1234
+
+class KeysightN7776C(Instrument):
+    """
+    This represents the Keysight N7776C Tunable Laser Source interface.
+    .. code-block:: python
+        laser = N7776C( address )
+        laser.sweep_wl_start = 1550
+        laser.sweep_wl_stop = 1560
+        laser.sweep_speed = 1
+        laser.sweep_mode = 'CONT'
+        laser.output = 1
+        while laser.sweep == 1:
+            continue
+        laser.output = 0
+    
+    """
+    def __init__(self, address, **kwargs):
+        super(KeysightN7776C, self).__init__(
+            address, "N7776C Tunable Laser Source",**kwargs)
+
+    def reset(self):
+        """
+        Reset the laser controller to its default state.
+        """
+        self.write('*RST')
+
+    def set_power(self,value,unit='mW'):
+        """
+        Function wrapper to set the power in an arbitrary unit
+        """
+        if not unit in ['dBm','mW']:
+            raise ValueError('Unknown Power Unit.')
+
+        self._output_power_unit = unit
+        if unit == 'mW':
+            self._output_power = value*1e-3 #Conversion to mW
+        else:
+            self._output_power = value
+
+    def get_power(self,unit_output=False):
+        """
+        Function wrapper to get the power including the unit directly if neccessary.
+        """
+        power_reading = self._output_power
+        power_unit = self._output_power_unit
+        if power_unit == 'mW':
+            if unit_output:
+                return (power_reading*1e3,power_unit)
+            else:
+                return power_reading*1e3
+        else:
+            if unit_output:
+                return (power_reading,power_unit)
+            else:
+                return power_reading
```

Comment:
I'm not sure, this unit handling feels a bit clunky, do you really want to change the return value type depending on the unit_output? I'd prefer to wait for #248 and handle this properly, then.

Reply:
Also, same conversion confusion as above. Maybe I haven't had enough coffee?

Reply:
As you have noticed, passing multiple objects to/from properties starts to feel strange quickly. 
Maybe while you wait for #248, for now instead of the current get_power/set_power/_output_power define 2 properties `output_power_dBm` and `output_power_mW`, and have the user use floats when getting/setting with those -- the property name makes the unit clear. 
You can look at other instruments to check how to manually implement properties with validators etc. Our property factory will not be suitable as you have the additional complication of the unit setting/getting. If you don't get/set power at a high rate, or your instrument is fast, you can get/set the unit every time. If not, you might cache it in an attribute (e.g. `self._current_unit` to avoid at least some redundant communication.

When #248 is available (could take time, though), you can add a generic unitful `output_power`. Btw please make a short note in that issue that this part of your instrument would be a candidate for test-implementing a unitful property.

Reply:
I finally got around to testing this, generally I like they way of having two properties ´´´output_power_dBm´´´ and ´´´output_power_mW´´´. However there was a flaw in the logic: For setting all is fine, but getting is a problem as the value given back is always in the unit of ´´´_output_power_unit´´´. 

Specifically the following script does pose problems:
´´´
l._output_power_unit #Response: 'mW'
l.output_power_dBm = 7.2
l.output_power_dBm #Resonse: 0.00524807442
l._output_power_unit = 'dBm'
l.output_power_dBm #Resonse: 7.2
´´´

I tried to fix it by writing a function which comapres the desired unit to what the current unit is and converts if neccessary (which I tried to hide) and then passing this for the get_process, but apparently I have no access to the base class in the get_process function. I need to look deeper into this, but if you have any suggestions of how to make this work, let me know. 

For now, I commited this version with two options under 87a8437. If you have a suggestion or a hint if this could be a way, let me know. I will think about it in more detail too

Reply:
Found out how to fix it with decorators!